rootio/wbuffer.go

```diff
@@ -4,6 +4,159 @@
 
 package rootio
 
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+type wbuff struct {
+	p []byte // buffer of data to write on
+	c int    // current position in buffer of data
+}
+
+func (w *wbuff) Write(p []byte) (int, error) {
+	if w.c >= len(w.p) {
+		return 0, io.EOF
+	}
+	n := copy(w.p[w.c:], p)
+	w.c += n
+	return n, nil
+}
+
+func (w *wbuff) WriteBufByte(p byte) error {
+	if w.c >= len(w.p) {
+		return io.EOF
+	}
+	w.p[w.c] = p
+	w.c++
+	return nil
+}
+
 // WBuffer is a write-only ROOT buffer for streaming.
 type WBuffer struct {
+	w      *wbuff
+	err    error
+	offset uint32
+	refs   map[int64]interface{}
+	sictx  StreamerInfoContext
+}
+
+func NewWBuffer(data []byte, refs map[int64]interface{}, offset uint32, ctx StreamerInfoContext) *WBuffer {
+	if refs == nil {
+		refs = make(map[int64]interface{})
+	}
+
+	return &WBuffer{
+		w:      &wbuff{p: data, c: 0},
+		refs:   refs,
+		offset: offset,
+		sictx:  ctx,
+	}
+}
+
+func (w *WBuffer) WriteI8(i int8) {
+	if w.err != nil {
+		fmt.Errorf("error in WriteI8")
+	}
+	v := byte(i)
+	w.w.WriteBufByte(v)
+}
+
+func (w *WBuffer) WriteI16(i int16) {
+	if w.err != nil {
+		fmt.Errorf("error in WriteI16")
+	}
+	err := binary.Write(w.w, binary.BigEndian, i)
```

Comment:
this is fine as a first pass.
but it's not very efficiency-friendly as `binary.Write` will go through a bunch of `reflect` code (and reflection is slow) to write this `int16`.

a more efficient implementation would be to use `binary.ByteOrder.PutUint16` and the `wbuff` field of `WBuffer`.
(so we need to directly address its underlying buffer and make sure there's enough room.)

`binary.BigEndian` implements `binary.ByteOrder`.

Reply:
This remark is just for WriteI16 or is it for all the write functions?